upgrade/upgrades: TestRoleMembersIDMigration1500Users failed

upgrade/upgrades.TestRoleMembersIDMigration1500Users [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9342487?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9342487?buildTab=artifacts#/) on release-23.1 @ [43847a23a3277ef6d97a07a70a54310542058f63](https://github.com/cockroachdb/cockroach/commits/43847a23a3277ef6d97a07a70a54310542058f63):


```
I230330 15:47:24.484070 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4825  job 852363921689313281: updated with schema change for table 33, mutation 0
I230330 15:47:24.681196 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4826  job 852363922193645569: updated with schema change for table 23, mutation 0
I230330 15:47:24.688257 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4827  job 852363921684299777: updated with schema change for table 4, mutation 0
I230330 15:47:24.690079 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4828  job 852363921689313281: updated with schema change for table 33, mutation 0
I230330 15:47:24.905829 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4829  job 852363922193645569: updated with schema change for table 23, mutation 0
I230330 15:47:24.911768 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4830  job 852363921684299777: updated with schema change for table 4, mutation 0
I230330 15:47:24.912507 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4831  job 852363921689313281: updated with schema change for table 33, mutation 0
I230330 15:47:25.096695 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4832  job 852363922193645569: updated with schema change for table 23, mutation 0
I230330 15:47:25.101694 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4833  job 852363921684299777: updated with schema change for table 4, mutation 0
I230330 15:47:25.102648 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4834  job 852363921689313281: updated with schema change for table 33, mutation 0
I230330 15:47:25.354503 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4835  job 852363922193645569: updated with schema change for table 23, mutation 0
I230330 15:47:25.360293 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4836  job 852363921684299777: updated with schema change for table 4, mutation 0
I230330 15:47:25.364886 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4837  job 852363921689313281: updated with schema change for table 33, mutation 0
I230330 15:47:25.550174 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4838  job 852363922193645569: updated with schema change for table 23, mutation 0
I230330 15:47:25.569167 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4839  job 852363921684299777: updated with schema change for table 4, mutation 0
I230330 15:47:25.569964 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4840  job 852363921689313281: updated with schema change for table 33, mutation 0
I230330 15:47:25.788195 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4841  job 852363922193645569: updated with schema change for table 23, mutation 0
I230330 15:47:25.794441 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4842  job 852363921684299777: updated with schema change for table 4, mutation 0
I230330 15:47:25.795751 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4843  job 852363921689313281: updated with schema change for table 33, mutation 0
I230330 15:47:26.005394 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4844  job 852363922193645569: updated with schema change for table 23, mutation 0
I230330 15:47:26.009891 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4845  job 852363921684299777: updated with schema change for table 4, mutation 0
I230330 15:47:26.010598 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4846  job 852363921689313281: updated with schema change for table 33, mutation 0
I230330 15:47:26.236080 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4847  job 852363922193645569: updated with schema change for table 23, mutation 0
I230330 15:47:26.248323 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4848  job 852363921684299777: updated with schema change for table 4, mutation 0
I230330 15:47:26.251576 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4849  job 852363921689313281: updated with schema change for table 33, mutation 0
I230330 15:47:26.467180 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4850  job 852363922193645569: updated with schema change for table 23, mutation 0
I230330 15:47:26.475222 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4851  job 852363921684299777: updated with schema change for table 4, mutation 0
I230330 15:47:26.476017 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4852  job 852363921689313281: updated with schema change for table 33, mutation 0
I230330 15:47:26.681127 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4853  job 852363922193645569: updated with schema change for table 23, mutation 0
I230330 15:47:26.685709 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4854  job 852363921684299777: updated with schema change for table 4, mutation 0
I230330 15:47:26.686504 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4855  job 852363921689313281: updated with schema change for table 33, mutation 0
I230330 15:47:26.883089 7870 sql/table.go:224  [T1,n1,client=127.0.0.1:46048,user=root] 4856  job 852363922193645569: updated with schema change for table 23, mutation 0
    helpers_test.go:190: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/upgrade/upgrades/helpers_test.go:190
        	            				github.com/cockroachdb/cockroach/pkg/upgrade/upgrades_test/pkg/upgrade/upgrades/role_members_ids_migration_test.go:77
        	            				github.com/cockroachdb/cockroach/pkg/upgrade/upgrades_test/pkg/upgrade/upgrades/role_members_ids_migration_test.go:41
        	Error:      	Received unexpected error:
        	            	pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh - unknown error: <nil>): "sql txn" meta={id=7e03f8d4 key=/Table/4/1/"testuser800"/0 pri=0.01049158 epo=0 ts=1680191243.714324823,2 min=1680191225.704495239,0 seq=1021} lock=true stat=PENDING rts=1680191225.704495239,0 wto=false gul=1680191226.204495239,0
        	Test:       	TestRoleMembersIDMigration1500Users
I230330 15:47:26.897768 6603 testutils/testcluster/testcluster.go:144  [-] 4857  TestCluster quiescing nodes
E230330 15:47:26.898088 27488 kv/kvserver/queue.go:1142  [T1,n1,raftlog,s1,r8/1:/Table/{4-5}] 4858  failed to send RPC: sending to all replicas failed; last error: ba: TruncateLog [/Table/4,/Min) RPC error: node unavailable; try another peer
W230330 15:47:26.898100 27487 kv/kvserver/intentresolver/intent_resolver.go:614  [-] 4859  failed to cleanup transaction intents: failed to resolve intents: node unavailable; try another peer
I230330 15:47:26.898274 7283 sql/stats/automatic_stats.go:507  [T1,n1] 4860  quiescing auto stats refresher
I230330 15:47:26.898354 7893 rpc/context.go:2302  [T1,n1,rnode=1,raddr=127.0.0.1:41017,class=default,rpc] 4861  connection heartbeat loop ended with err: <nil>
I230330 15:47:26.898751 6870 server/start_listen.go:103  [T1,n1] 4862  server shutting down: instructing cmux to stop accepting
W230330 15:47:26.899181 7064 sql/sqlliveness/slinstance/slinstance.go:312  [n1] 4863  exiting heartbeat loop
W230330 15:47:26.899220 7064 sql/sqlliveness/slinstance/slinstance.go:297  [n1] 4864  exiting heartbeat loop
E230330 15:47:26.899235 7064 server/server_sql.go:505  [n1] 4865  failed to run update of instance with new session ID: node unavailable; try another peer
E230330 15:47:26.899398 7368 jobs/registry.go:924  [T1,n1] 4866  error getting live session: node unavailable; try another peer
--- FAIL: TestRoleMembersIDMigration1500Users (118.96s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-schema
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRoleMembersIDMigration1500Users.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/upgrade/upgrades/BUILD.bazel
+++ b/pkg/upgrade/upgrades/BUILD.bazel
@@ -188,6 +188,7 @@ go_test(
         "//pkg/util/log",
         "//pkg/util/syncutil",
         "//pkg/util/timeutil",
+        "@com_github_cockroachdb_cockroach_go_v2//crdb",
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_stretchr_testify//assert",
         "@com_github_stretchr_testify//require",

--- a/pkg/upgrade/upgrades/database_role_settings_table_user_id_migration_test.go
+++ b/pkg/upgrade/upgrades/database_role_settings_table_user_id_migration_test.go
@@ -12,6 +12,7 @@ package upgrades_test
 
 import (
 	"context"
+	gosql "database/sql"
 	"fmt"
 	"strconv"
 	"testing"
@@ -64,9 +65,14 @@ func runTestDatabaseRoleSettingsUserIDMigration(t *testing.T, numUsers int) {
 	tdb := sqlutils.MakeSQLRunner(db)
 
 	// Create test users and add rows for each user to system.database_role_settings.
-	upgrades.ExecForCountInTxns(ctx, t, db, numUsers, 100 /* txCount */, func(txRunner *sqlutils.SQLRunner, i int) {
-		txRunner.Exec(t, fmt.Sprintf("CREATE USER testuser%d", i))
-		txRunner.Exec(t, fmt.Sprintf(`ALTER USER testuser%d SET application_name = 'roach sql'`, i))
+	upgrades.ExecForCountInTxns(ctx, t, db, numUsers, 100 /* txCount */, func(tx *gosql.Tx, i int) error {
+		if _, err := tx.Exec(fmt.Sprintf("CREATE USER testuser%d", i)); err != nil {
+			return err
+		}
+		if _, err := tx.Exec(fmt.Sprintf(`ALTER USER testuser%d SET application_name = 'roach sql'`, i)); err != nil {
+			return err
+		}
+		return nil
 	})
 	tdb.CheckQueryResults(t, "SELECT count(*) FROM system.database_role_settings", [][]string{{strconv.Itoa(numUsers)}})
 

--- a/pkg/upgrade/upgrades/external_connections_table_user_id_migration_test.go
+++ b/pkg/upgrade/upgrades/external_connections_table_user_id_migration_test.go
@@ -12,6 +12,7 @@ package upgrades_test
 
 import (
 	"context"
+	gosql "database/sql"
 	"fmt"
 	"strconv"
 	"testing"
@@ -60,14 +61,24 @@ func runTestExternalConnectionsUserIDMigration(t *testing.T, numUsers int) {
 	tdb := sqlutils.MakeSQLRunner(db)
 
 	// Create test users.
-	upgrades.ExecForCountInTxns(ctx, t, db, numUsers, 100 /* txCount */, func(txRunner *sqlutils.SQLRunner, i int) {
-		txRunner.Exec(t, fmt.Sprintf("CREATE USER testuser%d", i))
-		txRunner.Exec(t, fmt.Sprintf("GRANT SYSTEM EXTERNALCONNECTION TO testuser%d", i))
+	upgrades.ExecForCountInTxns(ctx, t, db, numUsers, 100 /* txCount */, func(tx *gosql.Tx, i int) error {
+		if _, err := tx.Exec(fmt.Sprintf("CREATE USER testuser%d", i)); err != nil {
+			return err
+		}
+		if _, err := tx.Exec(fmt.Sprintf("GRANT SYSTEM EXTERNALCONNECTION TO testuser%d", i)); err != nil {
+			return err
+		}
+		return nil
 	})
-	upgrades.ExecForCountInTxns(ctx, t, db, numUsers, 100 /* txCount */, func(txRunner *sqlutils.SQLRunner, i int) {
+	upgrades.ExecForCountInTxns(ctx, t, db, numUsers, 100 /* txCount */, func(tx *gosql.Tx, i int) error {
+		if _, err := tx.Exec(fmt.Sprintf("SET ROLE TO testuser%d", i)); err != nil {
+			return err
+		}
 		externalConnName := fmt.Sprintf("connection%d", i)
-		txRunner.Exec(t, fmt.Sprintf("SET ROLE TO testuser%d", i))
-		txRunner.Exec(t, fmt.Sprintf("CREATE EXTERNAL CONNECTION %[1]s AS 'userfile:///%[1]s'", externalConnName))
+		if _, err := tx.Exec(fmt.Sprintf("CREATE EXTERNAL CONNECTION %[1]s AS 'userfile:///%[1]s'", externalConnName)); err != nil {
+			return err
+		}
+		return nil
 	})
 	tdb.Exec(t, "SET ROLE TO root")
 	tdb.CheckQueryResults(t, "SELECT count(*) FROM system.external_connections", [][]string{{strconv.Itoa(numUsers)}})

--- a/pkg/upgrade/upgrades/helpers_test.go
+++ b/pkg/upgrade/upgrades/helpers_test.go
@@ -13,8 +13,10 @@ package upgrades
 import (
 	"context"
 	gosql "database/sql"
+	"math"
 	"testing"
 
+	"github.com/cockroachdb/cockroach-go/v2/crdb"
 	"github.com/cockroachdb/cockroach/pkg/clusterversion"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
@@ -24,7 +26,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/tabledesc"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
@@ -178,22 +179,22 @@ func ExecForCountInTxns(
 	db *gosql.DB,
 	count int,
 	txnSize int,
-	fn func(txRunner *sqlutils.SQLRunner, i int),
+	fn func(tx *gosql.Tx, i int) error,
 ) {
-	tx, err := db.BeginTx(ctx, nil /* opts */)
-	require.NoError(t, err)
-	txRunner := sqlutils.MakeSQLRunner(tx)
-	// Group statements into transactions of txnSize runs to speed up creation.
-	for i := 0; i < count; i++ {
-		if i != 0 && i%txnSize == 0 {
-			err := tx.Commit()
-			require.NoError(t, err)
-			tx, err = db.BeginTx(ctx, nil /* opts */)
-			require.NoError(t, err)
-			txRunner = sqlutils.MakeSQLRunner(tx)
+	numTxns := int(math.Ceil(float64(count) / float64(txnSize)))
+	for txnNum := 0; txnNum < numTxns; txnNum++ {
+		iterEnd := (txnNum + 1) * txnSize
+		if count < iterEnd {
+			iterEnd = count
 		}
-		fn(txRunner, i)
+		err := crdb.ExecuteTx(ctx, db, nil /* opts */, func(tx *gosql.Tx) error {
+			for i := txnNum * txnSize; i < iterEnd; i++ {
+				if err := fn(tx, i); err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+		require.NoError(t, err)
 	}
-	err = tx.Commit()
-	require.NoError(t, err)
 }

--- a/pkg/upgrade/upgrades/role_members_ids_migration_test.go
+++ b/pkg/upgrade/upgrades/role_members_ids_migration_test.go
@@ -12,6 +12,7 @@ package upgrades_test
 
 import (
 	"context"
+	gosql "database/sql"
 	"fmt"
 	"math/rand"
 	"testing"
@@ -73,22 +74,28 @@ func runTestRoleMembersIDMigration(t *testing.T, numUsers int) {
 	})
 
 	// Create test users.
-	expectedNumRoleMembersRows := 1
-	upgrades.ExecForCountInTxns(ctx, t, db, numUsers, 100 /* txCount */, func(txRunner *sqlutils.SQLRunner, i int) {
-		txRunner.Exec(t, fmt.Sprintf("CREATE USER testuser%d", i))
+	upgrades.ExecForCountInTxns(ctx, t, db, numUsers, 100 /* txCount */, func(tx *gosql.Tx, i int) error {
+		if _, err := tx.Exec(fmt.Sprintf("CREATE USER testuser%d", i)); err != nil {
+			return err
+		}
 		if i == 0 {
-			return
+			if _, err := tx.Exec(fmt.Sprintf("GRANT admin TO testuser%d", i)); err != nil {
+				return err
+			}
+			return nil
 		}
 		// Randomly choose an earlier test user to grant to the current test user.
-		grantStmt := fmt.Sprintf("GRANT testuser%d to testuser%d", rand.Intn(i), i)
+		grantStmt := fmt.Sprintf("GRANT testuser%d TO testuser%d", rand.Intn(i), i)
 		if rand.Intn(2) == 1 {
 			grantStmt += " WITH ADMIN OPTION"
 		}
-		txRunner.Exec(t, grantStmt)
-		expectedNumRoleMembersRows += 1
+		if _, err := tx.Exec(grantStmt); err != nil {
+			return err
+		}
+		return nil
 	})
 	tdb.CheckQueryResults(t, "SELECT count(*) FROM system.role_members", [][]string{
-		{fmt.Sprintf("%d", expectedNumRoleMembersRows)},
+		{fmt.Sprintf("%d", numUsers+1)},
 	})
 
 	// Run migrations.
@@ -104,7 +111,7 @@ func runTestRoleMembersIDMigration(t *testing.T, numUsers int) {
 	tdb.CheckQueryResults(t, "SELECT count(*) FROM system.role_members AS a JOIN system.users AS b on a.role = b.username AND a.role_id <> b.user_id", [][]string{{"0"}})
 	tdb.CheckQueryResults(t, "SELECT count(*) FROM system.role_members AS a JOIN system.users AS b on a.member = b.username AND a.member_id <> b.user_id", [][]string{{"0"}})
 
-	tdb.CheckQueryResults(t, `SELECT * FROM system.role_members WHERE "role" = 'admin'`, [][]string{
+	tdb.CheckQueryResults(t, `SELECT * FROM system.role_members WHERE "role" = 'admin' AND "member" = 'root'`, [][]string{
 		{"admin", "root", "true", "2", "1"},
 	})
 

--- a/pkg/upgrade/upgrades/system_privileges_user_id_migration_test.go
+++ b/pkg/upgrade/upgrades/system_privileges_user_id_migration_test.go
@@ -12,6 +12,7 @@ package upgrades_test
 
 import (
 	"context"
+	gosql "database/sql"
 	"fmt"
 	"strconv"
 	"testing"
@@ -64,9 +65,14 @@ func runTestSystemPrivilegesUserIDMigration(t *testing.T, numUsers int) {
 	tdb := sqlutils.MakeSQLRunner(db)
 
 	// Create test users and add rows for each user to system.privileges.
-	upgrades.ExecForCountInTxns(ctx, t, db, numUsers, 100 /* txCount */, func(txRunner *sqlutils.SQLRunner, i int) {
-		txRunner.Exec(t, fmt.Sprintf("CREATE USER testuser%d", i))
-		txRunner.Exec(t, fmt.Sprintf("GRANT SYSTEM MODIFYCLUSTERSETTING TO testuser%d", i))
+	upgrades.ExecForCountInTxns(ctx, t, db, numUsers, 100 /* txCount */, func(tx *gosql.Tx, i int) error {
+		if _, err := tx.Exec(fmt.Sprintf("CREATE USER testuser%d", i)); err != nil {
+			return err
+		}
+		if _, err := tx.Exec(fmt.Sprintf("GRANT SYSTEM MODIFYCLUSTERSETTING TO testuser%d", i)); err != nil {
+			return err
+		}
+		return nil
 	})
 	tdb.CheckQueryResults(t, "SELECT count(*) FROM system.privileges", [][]string{{strconv.Itoa(numUsers)}})
 

--- a/pkg/upgrade/upgrades/web_sessions_table_user_id_migration_test.go
+++ b/pkg/upgrade/upgrades/web_sessions_table_user_id_migration_test.go
@@ -12,6 +12,7 @@ package upgrades_test
 
 import (
 	"context"
+	gosql "database/sql"
 	"fmt"
 	"strconv"
 	"testing"
@@ -63,11 +64,12 @@ func runTestWebSessionsUserIDMigration(t *testing.T, numUsers int) {
 	tdb := sqlutils.MakeSQLRunner(db)
 
 	// Create test users.
-	upgrades.ExecForCountInTxns(ctx, t, db, numUsers, 100 /* txCount */, func(txRunner *sqlutils.SQLRunner, i int) {
-		txRunner.Exec(t, fmt.Sprintf("CREATE USER testuser%d", i))
-
+	upgrades.ExecForCountInTxns(ctx, t, db, numUsers, 100 /* txCount */, func(tx *gosql.Tx, i int) error {
+		if _, err := tx.Exec(fmt.Sprintf("CREATE USER testuser%d", i)); err != nil {
+			return err
+		}
 		// Simulate the INSERT that happens in the actual authentication code.
-		txRunner.Exec(t, fmt.Sprintf(`
+		if _, err := tx.Exec(fmt.Sprintf(`
 INSERT INTO system.web_sessions ("hashedSecret", username, "createdAt", "expiresAt", "lastUsedAt")
 VALUES (
 	'\xe77edd369fc3a955a129d2ced69d97717eeee49912e4369a2b687a4d8c36f798',
@@ -76,7 +78,10 @@ VALUES (
 	'2023-02-21 20:56:30.699242',
 	'2023-02-14 20:56:30.699447'
 )
-`, i))
+`, i)); err != nil {
+			return err
+		}
+		return nil
 	})
 	tdb.CheckQueryResults(t, "SELECT count(*) FROM system.web_sessions", [][]string{{strconv.Itoa(numUsers)}})
 

